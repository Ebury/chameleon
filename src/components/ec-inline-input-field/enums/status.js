export const EDITING = 'editing';
export const LOADING = 'loading';
export const READ_ONLY = 'read-only';
