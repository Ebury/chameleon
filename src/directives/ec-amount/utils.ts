
import defaultOptions from './options';
import type { AmountDirectiveOptions } from './types';

function format(input: undefined, opts?: AmountDirectiveOptions): undefined;
function format(input: null, opts?: AmountDirectiveOptions): null;
function format(input: string | number, opts?: AmountDirectiveOptions): string;
function format(input: string | number | undefined | null, opts: AmountDirectiveOptions = defaultOptions): string | null | undefined {
  /* eslint-disable no-param-reassign */
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.startsWith('-') ? '-' : '';

  input = sanitizeInput(input, opts.decimalSeparator);
  if (opts.precision > 0) {
    input = keepOnlyOneDecimalSeparator(input, opts.decimalSeparator);
  } else {
    input = removeDecimal(input, opts.decimalSeparator);
  }
  const parts = input.split(opts.decimalSeparator);
  let integralPart = parseToStr(parts[0]);
  let fractionalPart = parts[1];

  if (fractionalPart && fractionalPart.length > opts.precision) {
    fractionalPart = fractionalPart.substring(0, opts.precision);
  }

  integralPart = addGroupingSeparators(integralPart, opts.groupingSeparator);
  const result = negative + joinIntegralAndFractionalParts(integralPart, fractionalPart, opts.decimalSeparator);
  return result;
}

function parseToStr(number: string): string {
  const parsed = Number.parseFloat(number);
  if (Number.isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number: string, decimalSeparator: string): string {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function removeDecimal(number: string, decimalSeparator: string): string {
  const split = number.split(decimalSeparator);
  return split[0];
}

function unFormat(input: string, groupingSeparator: string, decimalSeparator: string): string {
  if (!input) {
    return input;
  }

  return input
    .replace(new RegExp(`\\${groupingSeparator}`, 'g'), '')
    .replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
}

function sanitizeInput(input: string, decimalSeparator: string): string {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integralPart: string, separator: string): string {
  return integralPart.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegralAndFractionalParts(integralPart: string, fractionalPart: string, decimalSeparator: string): string {
  return fractionalPart !== undefined ? integralPart + decimalSeparator + fractionalPart : integralPart;
}

/* c8 ignore start */
function setCursor(el: HTMLInputElement, position: number) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}
/* c8 ignore stop */

export {
  format,
  sanitizeInput,
  setCursor,
  unFormat,
};
