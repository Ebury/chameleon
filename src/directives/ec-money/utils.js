import defaults from './options';

function format(input, opt = defaults) {
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.includes('-') ? '-' : '';

  input = sanitizeInput(input, opt.decimalSeparator);
  input = keepOnlyOneDecimalSeparator(input, opt.decimalSeparator);
  const parts = input.split(opt.decimalSeparator);
  let integer = parseToStr(parts[0]);
  let decimal = parts[1];

  if (decimal && decimal.length > opt.precision) {
    decimal = decimal.substr(0, opt.precision);
  }

  integer = addGroupingSeparators(integer, opt.groupingSeparator);
  const result = negative + joinIntegerAndDecimal(integer, decimal, opt.decimalSeparator);
  return result;
}

function parseToStr(number) {
  /* eslint no-restricted-globals: "off" */
  const parsed = parseFloat(number);
  if (isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number, decimalSeparator) {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function unFormat(input, groupingSeparator, decimalSeparator) {
  /* eslint no-param-reassign: "off" */
  input = (`${input}`).replace(new RegExp(`\\${groupingSeparator}`, 'g'), '');
  input = (`${input}`).replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
  return input;
}

function sanitizeInput(input, decimalSeparator) {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integer, separator) {
  return integer.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegerAndDecimal(integer, decimal, decimalSeparator) {
  return decimal !== undefined ? integer + decimalSeparator + decimal : integer;
}

function setCursor(el, position) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}

// https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events#The_old-fashioned_way
function event(name) {
  const evt = document.createEvent('Event');
  evt.initEvent(name, true, true);
  return evt;
}

export {
  format,
  unFormat,
  setCursor,
  event,
  sanitizeInput,
};
