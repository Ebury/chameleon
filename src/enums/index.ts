export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortDirectionCycle {
  LOWEST_FIRST = 1,
  HIGHEST_FIRST = 2,
}

export enum KeyCode {
  ESC = 27,
  ARROW_UP = 38,
  ARROW_DOWN = 40,
}

export enum ZIndexLevel {
  NOTIFICATION = 'notification',
  MODAL = 'modal',
  TOOLTIP = 'tooltip',
  LEVEL1 = 'level-1',
  LEVEL2 = 'level-2',
  LEVEL3 = 'level-3',
}
