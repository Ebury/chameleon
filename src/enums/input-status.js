export const EDIT = 'edit';
export const LOADING = 'loading';
export const READ_ONLY = 'read-only';
