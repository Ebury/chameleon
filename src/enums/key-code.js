export const ESC = 27;
