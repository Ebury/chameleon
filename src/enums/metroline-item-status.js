export const NEXT = 'next';
export const ACTIVE = 'active';
export const COMPLETED = 'completed';
