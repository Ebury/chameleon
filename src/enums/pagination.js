export const PAGE_SIZES = [5, 10, 50, 100];
export const DEFAULT_PAGE_SIZE = 10;
