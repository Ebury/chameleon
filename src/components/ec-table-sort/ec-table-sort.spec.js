import { mount } from '@vue/test-utils';

import { withMockedConsole } from '../../../tests/utils/console';
import * as SortDirection from '../../enums/sort-direction';
import EcTableSort from './ec-table-sort.vue';

describe('EcTableSort', () => {
  function mountEcTableSort(props, mountOpts) {
    return mount(EcTableSort, {
      props,
      ...mountOpts,
    });
  }

  it('should render as expected', () => {
    const wrapper = mountEcTableSort();
    expect(wrapper.element).toMatchSnapshot();
  });

  describe(':props', () => {
    it('should render with no direction', () => {
      const wrapper = mountEcTableSort({ direction: null });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render with direction set to empty string', () => {
      const wrapper = mountEcTableSort({ direction: '' });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render with direction set to ASC', () => {
      const wrapper = mountEcTableSort({ direction: SortDirection.ASC });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render with direction set to DESC', () => {
      const wrapper = mountEcTableSort({ direction: SortDirection.DESC });
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should validate direction prop', () => {
      withMockedConsole((errorSpy, warnSpy) => {
        mountEcTableSort({ direction: 'invalid' });
        expect(warnSpy).toHaveBeenCalledTimes(1);
        expect(warnSpy.mock.calls[0][0]).toContain('Invalid prop: custom validator check failed for prop "direction"');
      });
    });
  });

  describe('@events', () => {
    it('should notify parent about sorting', () => {
      const wrapper = mountEcTableSort({ direction: SortDirection.ASC });

      wrapper.findByDataTest('ec-table-sort__icon').trigger('click');
      expect(wrapper.emitted('sort')).toEqual([[SortDirection.ASC]]);
    });
  });

  describe('#slots', () => {
    it('should render default slot', () => {
      const wrapper = mountEcTableSort({}, {
        slots: {
          default: '<div>Random title</div>',
        },
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
