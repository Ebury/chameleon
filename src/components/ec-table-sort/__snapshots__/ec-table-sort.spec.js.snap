// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcTableSort #slots should render default slot 1`] = `
<a
  class="ec-table-sort"
  data-test="ec-table-sort"
  href="#"
  title="Not sorted"
>
  <svg
    class="ec-icon ec-table-sort__icon"
    data-test="ec-table-sort__icon"
    height="16"
    width="16"
  >
    <use
      href="/img/simple-icons.svg#ec-simple-arrow-up-down"
    />
  </svg>
   
  <!---->
</a>
`;

exports[`EcTableSort :props should render with direction set to ASC 1`] = `
<a
  class="ec-table-sort"
  data-test="ec-table-sort"
  href="#"
  title="Sorted ascending"
>
  <svg
    class="ec-icon ec-table-sort__icon ec-table-sort__icon--asc"
    data-test="ec-table-sort__icon"
    height="16"
    width="16"
  >
    <use
      href="/img/simple-icons.svg#ec-simple-arrow-drop-down"
    />
  </svg>
   
  <span
    class="ec-table-sort__text"
  >
    asc
  </span>
</a>
`;

exports[`EcTableSort :props should render with direction set to DESC 1`] = `
<a
  class="ec-table-sort"
  data-test="ec-table-sort"
  href="#"
  title="Sorted descending"
>
  <svg
    class="ec-icon ec-table-sort__icon ec-table-sort__icon--desc"
    data-test="ec-table-sort__icon"
    height="16"
    width="16"
  >
    <use
      href="/img/simple-icons.svg#ec-simple-arrow-drop-down"
    />
  </svg>
   
  <span
    class="ec-table-sort__text"
  >
    desc
  </span>
</a>
`;

exports[`EcTableSort :props should render with direction set to empty string 1`] = `
<a
  class="ec-table-sort"
  data-test="ec-table-sort"
  href="#"
  title="Not sorted"
>
  <svg
    class="ec-icon ec-table-sort__icon"
    data-test="ec-table-sort__icon"
    height="16"
    width="16"
  >
    <use
      href="/img/simple-icons.svg#ec-simple-arrow-up-down"
    />
  </svg>
   
  <!---->
</a>
`;

exports[`EcTableSort :props should render with no direction 1`] = `
<a
  class="ec-table-sort"
  data-test="ec-table-sort"
  href="#"
  title="Not sorted"
>
  <svg
    class="ec-icon ec-table-sort__icon"
    data-test="ec-table-sort__icon"
    height="16"
    width="16"
  >
    <use
      href="/img/simple-icons.svg#ec-simple-arrow-up-down"
    />
  </svg>
   
  <!---->
</a>
`;

exports[`EcTableSort should render as expected 1`] = `
<a
  class="ec-table-sort"
  data-test="ec-table-sort"
  href="#"
  title="Not sorted"
>
  <svg
    class="ec-icon ec-table-sort__icon"
    data-test="ec-table-sort__icon"
    height="16"
    width="16"
  >
    <use
      href="/img/simple-icons.svg#ec-simple-arrow-up-down"
    />
  </svg>
   
  <!---->
</a>
`;
